tests/ttnn/unit_tests/gtests/ccl/kernels/fabric_worker_sender_multi_input.cpp

```diff
@@ -0,0 +1,222 @@
+
+    // We need to normalize all noc addresses to be for a consistent noc ID
+    // so the remote sender core can correctly send the packet. In the future
+    // we can decide if it's better for the noc index to be embedded in the packet
+    // header (for now we don't do that)
+    constexpr size_t NORMALIZED_NOC_INDEX = 0;
+
+    cb_wait_front(cb_id_in0, 1);
+    auto a_packet_header_addr = get_read_ptr(cb_id_in0);
+
+    if constexpr (read_mode == ReadMode::FULLY_ORDERED || read_mode == ReadMode::RATIOD_FORWARDING) {
+
+        size_t current_page_in0 = 0;
+        size_t current_page_in1 = 0;
+        while (current_page_in0 < total_pages_to_send || current_page_in1 < total_pages_to_send) {
+            for (size_t read = 0; read < read_ratio0 && current_page_in0 < total_pages_to_send; current_page_in0 += num_pages_per_send, read++) {
+                uint32_t pages_to_send = std::min<uint32_t>(num_pages_per_send, total_pages_to_send - current_page_in0);
+                cb_wait_front(cb_id_in0, pages_to_send);
+                non_blocking_read_and_forward(current_page_in0, cb_id_in0, dest_addr_gen0, sender, config, page_size, total_pages_to_send, num_pages_per_send);
+                cb_pop_front(cb_id_in0, pages_to_send);
+            }
+
+            for (size_t read = 0; read < read_ratio1 && current_page_in1 < total_pages_to_send; current_page_in1 += num_pages_per_send, read++) {
+                uint32_t pages_to_send = std::min<uint32_t>(num_pages_per_send, total_pages_to_send - current_page_in1);
+                cb_wait_front(cb_id_in1, pages_to_send);
+                non_blocking_read_and_forward(current_page_in1, cb_id_in1, dest_addr_gen1, sender, config, page_size, total_pages_to_send, num_pages_per_send);
+                cb_pop_front(cb_id_in1, pages_to_send);
+            }
+        }
+
+    } else if constexpr (read_mode == ReadMode::ARBITRARILY_ORDERED) {
+        size_t current_page_in0 = 0;
+        size_t current_page_in1 = 0;
+        while (current_page_in0 < total_pages_to_send || current_page_in1 < total_pages_to_send) {
+            if (current_page_in0 < total_pages_to_send) {
+                non_blocking_read_and_forward(current_page_in0, cb_id_in0, dest_addr_gen0, sender, config, page_size, total_pages_to_send, num_pages_per_send);
+            }
+            if (current_page_in1 < total_pages_to_send) {
+                non_blocking_read_and_forward(current_page_in1, cb_id_in1, dest_addr_gen1, sender, config, page_size, total_pages_to_send, num_pages_per_send);
+            }
+        }
+    }
+
+    sender.wait_for_empty_write_slot();
+
+    auto &packet_header = *reinterpret_cast<tt::fabric::PacketHeader*>(a_packet_header_addr);
+    ASSERT(*last_message_semaphore_address == 0);
+    packet_header.reserved = 0xE;
+    packet_header.reserved2 = 0xFFFF;
+    packet_header.to_atomic_inc();
+    packet_header.to_chip_unicast(tt::fabric::UnicastRoutingCommandHeader{2});
```

Comment:
Same comment as earlier, are the 0xE and 2 just arbitrary values for the handshake or do they mean something